### PR TITLE
Fixes #3 broken PDF thumnail generation

### LINKED ISSUE #3: PDF thumbnails not being generated
#### Current Behavior
The PDF thumbnails aren't generated on upload, and instead falling back to a generic PDF icon.
Screenshot from 2020-09-17 16-58-11

#### Steps to Reproduce
Upload a PDF, and notice that the thumbnail never gets generated.

#### Expected Behavior
On dev and prod, PDF thumbnails are successfully generated after upload.

#### Possible Solutions
Found this error in the logs:

```
[12-Sep-2020 15:06:09 America/Chicago] PID 8217 - /data/www/task-runner/action/thumbnail/thumbnail.action.php -> UploadThumbToS3 at line 851 - Mem 10.08 - 777.77ms -
 Message: 
object(stdClass)#23 (5) {
  ["__CLASS__"]=>  string(16) "ImagickException"
  ["message"]=>  string(115) "attempt to perform an operation not allowed by the security policy `PDF' @ error/constitute.c/IsCoderAuthorized/408"
  ["code"]=>  int(499)
  ["file"]=>  string(36) "/data/www/library/util/thumbnail.php"
  ["line"]=>  int(17)
}
```
This points to an `ImageMagick` policy issue. The solution is to use the current policy on prod, since this works there, and check that in to this repo.

### PR DESCRIPTION
**Pull request by xmunoz**
- Make `SQS_IDENT` and `DATA_DELETE` required environment variables
- Run the provisioning script every time to implement latest changes on `vagrant up`
- Add environment template file for keeping track of required environment variables
- Add `policy.xml` to repo
- Update README with new workflow instructions

## Steps to Test
- Read the README
- `vagrant halt`
- `vagrant up`

### CODE REVIEW

File: vagrant/etc/ImageMagick-6/policy.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE policymap [
+<!ELEMENT policymap (policy)+>
+<!ELEMENT policy (#PCDATA)>
+<!ATTLIST policy domain (delegate|coder|filter|path|resource) #IMPLIED>
+<!ATTLIST policy name CDATA #IMPLIED>
+<!ATTLIST policy rights CDATA #IMPLIED>
+<!ATTLIST policy pattern CDATA #IMPLIED>
+<!ATTLIST policy value CDATA #IMPLIED>
+]>
```
Comment (jasonaowen) [inline vagrant/etc/ImageMagick-6/policy.xml:10]: Where did this `<!DOCTYPE>` declaration come from? It doesn't match the one listed at https://imagemagick.org/source/policy.xml, linked to from https://imagemagick.org/script/security-policy.php .

File: vagrant/etc/ImageMagick-6/policy.xml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+<policymap>
+  <policy domain="coder" rights="read|write" pattern="EPHEMERAL" />
+  <policy domain="coder" rights="read|write" pattern="URL" />
+  <policy domain="coder" rights="read|write" pattern="HTTPS" />
```
Comment (jasonaowen) [inline vagrant/etc/ImageMagick-6/policy.xml:14]: I don't know what most of these mean, but granting read and write permissions to network resources doesn't sound great to me:

> If you utilize ImageMagick from a public website, you may want to increase security by preventing usage of the MVG or HTTPS coders.

It's wild that this is even a capability ImageMagick has built!

**Review by jasonaowen — COMMENTED** [review]
Are the SQS changes related to the ImageMagick changes? Would you mind splitting them out to separate commits?

Can we link to the [security policy documentation](https://imagemagick.org/script/security-policy.php), either in the commit message or as a comment in the XML?

Also, can you please write a bit more about how this solves the problem? If the problem is large files, I don't understand why this `policy.xml` fixes it; is it just removing the Ubuntu-provided limits and letting ImageMagick use as much memory and disk as it wants?

**Comment by jasonaowen** [comment]
> Also, can you please write a bit more about how this solves the problem? If the problem is large files, I don't understand why this `policy.xml` fixes it; is it just removing the Ubuntu-provided limits and letting ImageMagick use as much memory and disk as it wants?

Ah, I hadn't read the post-edit issue #3. The problem is that the default Ubuntu security policy prevents ImageMagick from operating on PDF files, due to [a previous security vulnerability](https://www.kb.cert.org/vuls/id/332928), and the problem is solved by replacing it with a `policy.xml` that does not contain that restriction against PDFs.

Do I have that right?

File: vagrant/etc/ImageMagick-6/policy.xml (hunk 1 above)
Comment (xmunoz) [reply vagrant/etc/ImageMagick-6/policy.xml:10]: No idea!

**Comment by xmunoz** [comment]
I believe so. We can discuss more offline. Unfortunately, this is the `policy.xml` that we must use for now. I will create a new JIRA issue to keep track of further work needed here.

> Would you mind splitting them out to separate commits?
That is something that I can, and will do now.

File: Vagrantfile [hunk 3]
```diff
@@ -67,7 +66,9 @@ Vagrant.configure(2) do |config|
   config.vm.provision "shell", path: "bootstrap.sh",
     env: {"AWS_ACCESS_KEY_ID": ENV["AWS_ACCESS_KEY_ID"],
           "AWS_ACCESS_SECRET": ENV["AWS_ACCESS_SECRET"],
-          "AWS_REGION": ENV["AWS_REGION"]}
+          "AWS_REGION": ENV["AWS_REGION"],
+          "SQS_IDENT": ENV["SQS_IDENT"]},
+    run: "always"
```
Comment (jasonaowen) [inline Vagrantfile:71]: If we run `bootstrap.sh` every time, does that mean the database gets wiped and the S3 bucket gets emptied every time we `vagrant up`, even after a `vagrant suspend`? That seems... not great.

Comment (xmunoz) [reply Vagrantfile:71]: It's not great. I'll add some logic in the bash script that asks you if you want to wipe stateful data.

File: README.md [hunk 4]
```diff
@@ -28,17 +28,13 @@ If this command fails, check out [Troubleshooting](#troubleshooting) for suggest
 - Local_Video_YourName
 - Local_High_Priority_YourName
 
-4. Export AWS configuration variables.
+4. Define environment variables in `.env` file.
```
Comment (jasonaowen) [inline README.md:31]: This is great, and actually what I'm already doing!

Unfortunately, that means that checking out this branch conflicts with my local file, and will conflict with any local changes in the future, which these instructions are encouraging.

Instead, the pattern I've successfully used before is to check in a `.env.template` file with the relevant variables filled with example data, and have the instructions include a step like `cp .env.template .env`. You already made the template, and you've already added `.env` to the `.gitignore`, so all that needs to change is the template name and the instructions.

Here's a draft of the new instructions, in the hopes that it is helpful:

```suggestion
4. Define environment variables in `.env` file. Copy the template file (`cp .env.template .env`), and open the copy in your preferred editor.
```

**Review by jasonaowen — CHANGES_REQUESTED** [review]
+1 to documenting the `.env` pattern, but see below.

However, all of this is unrelated to the intended purpose of the pull request: fixing PDF thumbnail generation in the dev environment. Can we split the bootstrap changes out into a separate pull request, and reduce this PR to the single commit with the PDF changes? (Or vice versa, move the PDF commit to a new branch + PR).

**Comment by xmunoz** [comment]
> However, all of this is unrelated to the intended purpose of the pull request: fixing PDF thumbnail generation in the dev environment. Can we split the bootstrap changes out into a separate pull request, and reduce this PR to the single commit with the PDF changes? (Or vice versa, move the PDF commit to a new branch + PR).

No, I don't want to do that. For better or worse, this PR contains all of the review comments and context. I will ensure the PR description captures all of the work that happened here, but I'm not going to open another PR unless there is a compelling technical reason to.

**Comment by jasonaowen** [comment]
My motivation in suggesting a split is to help get the PDF work in more quickly, which I think is ready to go, while giving us more space to discuss the bootstrap changes, which I haven't finished thinking through or testing. I'm working on that now, but as always the problem with rebuilding VMs is the time it takes to do so.

**Comment by jasonaowen** [comment]
I didn't fully think through the new `DELETE_DATA` environment variable before the meeting earlier today. I have some concerns about the approach, as I worry it makes it easy both to accidentally delete data you wanted to keep, and to spin up a new VM that is missing necessary data. The latter is more important, probably; "make errors inexpressible"!

I think what we want is to maintain the behavior of the previous data-destructive operation: run it when the VM is created, do not run it when an existing VM is brought back up after being suspended. Can we split out the relevant bits to a new script that is run with the same trigger as the bootstrap script previously had? Or, are there any vagrant-provided environment variables that we can examine to switch on?

**Comment by xmunoz** [comment]
> Or, are there any vagrant-provided environment variables that we can examine to switch on?
There are! If you pass the `--provision` to vagrant up, it will run the `bootstrap.sh` script. How about this?
1. I remove `run: "always"` from the shell provisioner.
2. Update the README to reference this `--provision` flag with a link to the [docs](https://www.vagrantup.com/docs/cli/up#no-provision), and clarify when and how to explicitly run the provisioner.

**Comment by jasonaowen** [comment]
Yes, that sounds like a great approach, @xmunoz ! Thanks.

**Comment by jasonaowen** [comment]
This might be an additional helpful doc link, [Provisioning](https://www.vagrantup.com/docs/provisioning), as it contains this bit that helped clarify my understanding:

> Provisioning happens ... [on] the first `vagrant up` that creates the environment... If the environment was already created and the up is just resuming a machine or booting it up, [provisioner scripts] will not run unless the `--provision` flag is explicitly provided.

**Review by jasonaowen — COMMENTED** [review]
Ugh, my local environment is broken again. I don't know if it's anything to do with these changes; I feel like it's probably not, but I'm out of steam for debugging this today. I'll get back to it tomorrow.

File: bootstrap.sh [hunk 5]
```diff
@@ -73,11 +75,14 @@ a2enconf charset
...
 service apache2 start
 
-echo "Populate MySQL"
-service mysql restart
-sudo mysql < /data/www/docker/database/perm.sql
-sudo mysql < /data/www/website/database/wp.sql
-sudo mysql wp < /data/www/website/database/dump.sql
+if $DELETE_DATA
+then
+    echo "Populate MySQL"
```
[the remark this answers is not in the document]
Comment (xmunoz) [reply bootstrap.sh:80]: TBH I don't understand the concern here. In the template, `DELETE_DATA` is set to the default correct value for the first provision attempt. The instructions are clear about possible values. There is also a pretty good change this entire script will be replaced by an Ansible playbook next week.

[the remark this answers is not in the document]
Comment (jasonaowen) [reply bootstrap.sh:80]: I'm concerned that this change introduces an opportunity for user error that did not exist before: setting `DELETE_DATA=false`, running `vagrant destroy; vagrant up` will result in a broken environment. I haven't yet tested what that looks like, but I am not optimistic that the cause of the problem will be immediately obvious, given the way other others manifest in this application.

I hear you and agree that the instructions are good, and that the chance of the error I'm describing is not high. If, however, we could reasonably engineer a solution where the chance of error is zero, that would be better. You know the tools involved much better than I do, so if say there isn't then I believe you, and am just a little bit sad.

[the remark this answers is not in the document]
Comment (xmunoz) [reply bootstrap.sh:81]: Yeah, I hear that, and it's definitely something that needs to be re-worked. Are you fine with this as an interim solution?

File: bootstrap.sh [hunk 6]
```diff
@@ -12,29 +12,29 @@ apt-get -qq update
...
 fi
 if [[ -z "$AWS_REGION" || -z "$AWS_ACCESS_KEY_ID" || -z "$AWS_ACCESS_SECRET" ]]
     then
 	echo "ERROR! Missing critical environment variable: AWS_REGION, AWS_ACCESS_KEY_ID, AWS_ACCESS_SECRET!"
     exit 1
 fi
-mv /home/vagrant/sqs.txt /data/www
+
 echo "local" > /data/www/host.txt
+runuser -l vagrant -c "echo $SQS_IDENT > /data/www/sqs.txt"
```
Comment (jasonaowen) [inline bootstrap.sh:28]: Is this going to do what we want? From `man runuser`:

```quote
       -, -l, --login
              Start the shell as a login shell with an environment similar to a real login:

                 o      clears all the environment variables except for TERM

                 o      initializes the environment variables HOME, SHELL, USER, LOGNAME, PATH

                 o      changes to the target user's home directory

                 o      sets argv[0] of the shell to '-' in order to make the shell a login shell
```

That "clears all the environment variables" bit suggests to me that `$SQS_IDENT` will not survive.

I think probably we want this to match the previous line:

```suggestion
echo $SQS_IDENT > /data/www/sqs.txt
```

I know you're in the process of `vagrant destroy; vagrant up`, so if it works for you as-is then great, no change needed!

File: bootstrap.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply bootstrap.sh:81]: Yes! Sorry, I should have made that more clear up front.

File: bootstrap.sh (hunk 6 above)
Comment (xmunoz) [reply bootstrap.sh]: This was bad for another reason too: this directory is only writeable by root at this point. This is the correct solution, which I realized independently and pushed. Thanks for catching it too though!~

**Comment by jasonaowen** [comment]
I'm spinning up a new instance based on the latest commit.

Meanwhile, the last thing I'm going to ask is that we clean up the commit history on this branch, so that what we merge is more communicative to future developers; our debugging process, as expressed through a series of fixup commits, is not interesting or informative.

**Review by jasonaowen — APPROVED** [review]
Woo! Tested successfully by creating a new VM with `vagrant destroy; vagrant up`, creating an account, uploading a file, and seeing it generate a thumbnail.

This looks good to me! Thanks for pairing with me on diagnosing problems and cleaning up commit history, @xmunoz!

**Pull request merged**